Fix for unhandled exceptions in the Triggerer tests

drajguru opened the pull request:
Fix for unhandled exceptions raised in the triggerer tests on the queue listener thread.

The queue listener was being killed due to missing a reference to the triggerer_job. Fixing this resulted in console logging during tests. So changes also include disabling the console logging handler / queue listener during tests.

Also fixed a bug in tests/jobs/test_triggerer_job.py::test_trigger_create_race_condition_18392 which caused another unhandled exception.

----

```console
tests/jobs/test_triggerer_job.py::test_trigger_logging_sensitive_info                                                                                                               
  /usr/local/lib/python3.8/site-packages/_pytest/threadexception.py:73: PytestUnhandledThreadExceptionWarning: Exception in thread Thread-3                                         
                                                                                                                                                                                    
  Traceback (most recent call last):                                                                                                                                                
    File "/usr/local/lib/python3.8/threading.py", line 932, in _bootstrap_inner                                                                                                     
      self.run()                                                                                                                                                                    
    File "/usr/local/lib/python3.8/threading.py", line 870, in run                                                                                                                  
      self._target(*self._args, **self._kwargs)                                                                                                                                     
    File "/usr/local/lib/python3.8/logging/handlers.py", line 1487, in _monitor                                                                                                     
      self.handle(record)                                                                                                                                                           
    File "/usr/local/lib/python3.8/logging/handlers.py", line 1468, in handle                                                                                                       
      handler.handle(record)                                                                                                                                                        
    File "/opt/airflow/airflow/utils/log/trigger_handler.py", line 107, in handle                                                                                                   
      self.emit(record)                                                                                                                                                             
    File "/opt/airflow/airflow/utils/log/trigger_handler.py", line 96, in emit                                                                                                      
      h = self._get_or_create_handler(record.trigger_id, record.task_instance)                                                                                                      
    File "/opt/airflow/airflow/utils/log/trigger_handler.py", line 92, in _get_or_create_handler                                                                                    
      self.handlers[trigger_id] = self._make_handler(ti)                                                                                                                            
    File "/opt/airflow/airflow/utils/log/trigger_handler.py", line 87, in _make_handler                                                                                             
      h.set_context(ti=ti)                                                                                                                                                          
    File "/opt/airflow/airflow/utils/log/file_task_handler.py", line 243, in set_context                                                                                            
      local_loc = self._init_file(ti, identifier=identifier)                                                                                                                        
    File "/opt/airflow/airflow/utils/log/file_task_handler.py", line 533, in _init_file                                                                                             
      full_path = self.add_triggerer_suffix(full_path=full_path, job_id=ti.triggerer_job.id)                                                                                        
  AttributeError: 'NoneType' object has no attribute 'id'                                                                                                                           
                                                                                                                                                                                    
    warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))                                                                                                                
                                                                                                                                                                                    
tests/jobs/test_triggerer_job.py::test_trigger_lifecycle                                                                                                                            
  /usr/local/lib/python3.8/site-packages/_pytest/threadexception.py:73: PytestUnhandledThreadExceptionWarning: Exception in thread Thread-11                                        
                                                                                                                                                                                    
  Traceback (most recent call last):                                                                                                                                                
    File "/usr/local/lib/python3.8/threading.py", line 932, in _bootstrap_inner                                                                                                     
      self.run()                                                                                                                                                                    
    File "/usr/local/lib/python3.8/threading.py", line 870, in run
      self._target(*self._args, **self._kwargs)
    File "/usr/local/lib/python3.8/logging/handlers.py", line 1487, in _monitor
      self.handle(record)
    File "/usr/local/lib/python3.8/logging/handlers.py", line 1468, in handle
      handler.handle(record)
    File "/opt/airflow/airflow/utils/log/trigger_handler.py", line 107, in handle
      self.emit(record)
    File "/opt/airflow/airflow/utils/log/trigger_handler.py", line 96, in emit
      h = self._get_or_create_handler(record.trigger_id, record.task_instance)
    File "/opt/airflow/airflow/utils/log/trigger_handler.py", line 92, in _get_or_create_handler
      self.handlers[trigger_id] = self._make_handler(ti)
    File "/opt/airflow/airflow/utils/log/trigger_handler.py", line 87, in _make_handler
      h.set_context(ti=ti)
    File "/opt/airflow/airflow/utils/log/file_task_handler.py", line 243, in set_context
      local_loc = self._init_file(ti, identifier=identifier)
    File "/opt/airflow/airflow/utils/log/file_task_handler.py", line 533, in _init_file
      full_path = self.add_triggerer_suffix(full_path=full_path, job_id=ti.triggerer_job.id)
  AttributeError: 'NoneType' object has no attribute 'id'


tests/jobs/test_triggerer_job.py::test_trigger_create_race_condition_18392
  /usr/local/lib/python3.8/site-packages/_pytest/threadexception.py:73: PytestUnhandledThreadExceptionWarning: Exception in thread Thread-19
   
  Traceback (most recent call last):
    File "/usr/local/lib/python3.8/threading.py", line 932, in _bootstrap_inner
      self.run()
    File "/usr/local/lib/python3.8/threading.py", line 870, in run
      self._target(*self._args, **self._kwargs)
    File "/opt/airflow/airflow/jobs/triggerer_job_runner.py", line 341, in _execute
      self._run_trigger_loop()
    File "/opt/airflow/airflow/jobs/triggerer_job_runner.py", line 364, in _run_trigger_loop
      self.load_triggers()
    File "/opt/airflow/tests/jobs/test_triggerer_job.py", line 376, in load_triggers
      self.wait_for_runner_loop(runner_loop_count=2)
    File "/opt/airflow/tests/jobs/test_triggerer_job.py", line 367, in wait_for_runner_loop
      pytest.fail("did not observe 2 loops in the runner thread")
    File "/usr/local/lib/python3.8/site-packages/_pytest/outcomes.py", line 198, in fail
      raise Failed(msg=reason, pytrace=pytrace)
  Failed: did not observe 2 loops in the runner thread
   
    warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))
                                                                                                              
                                                                                                                                                                         
tests/jobs/test_triggerer_job.py::test_trigger_firing                                                                                                                               
  /usr/local/lib/python3.8/site-packages/_pytest/threadexception.py:73: PytestUnhandledThreadExceptionWarning: Exception in thread Thread-28                                        
                                                                                                                                                                                    
  Traceback (most recent call last):                                                                                                                                                
    File "/usr/local/lib/python3.8/threading.py", line 932, in _bootstrap_inner                                                                                                     
      self.run()                                                                                                                                                                    
    File "/usr/local/lib/python3.8/threading.py", line 870, in run                                                                                                                  
      self._target(*self._args, **self._kwargs)                                                                                                                                     
    File "/usr/local/lib/python3.8/logging/handlers.py", line 1487, in _monitor                                                                                                     
      self.handle(record)                                                                                                                                                           
    File "/usr/local/lib/python3.8/logging/handlers.py", line 1468, in handle                                                                                                       
      handler.handle(record)                                                                                                                                                        
    File "/opt/airflow/airflow/utils/log/trigger_handler.py", line 107, in handle                                                                                                   
      self.emit(record)                                                                                                                                                             
    File "/opt/airflow/airflow/utils/log/trigger_handler.py", line 96, in emit                                                                                                      
      h = self._get_or_create_handler(record.trigger_id, record.task_instance)                                                                                                      
    File "/opt/airflow/airflow/utils/log/trigger_handler.py", line 92, in _get_or_create_handler                                                                                    
      self.handlers[trigger_id] = self._make_handler(ti)                                                                                                                            
    File "/opt/airflow/airflow/utils/log/trigger_handler.py", line 87, in _make_handler                                                                                             
      h.set_context(ti=ti)                                                                                                                                                          
    File "/opt/airflow/airflow/utils/log/file_task_handler.py", line 243, in set_context                                                                                            
      local_loc = self._init_file(ti, identifier=identifier)
    File "/opt/airflow/airflow/utils/log/file_task_handler.py", line 533, in _init_file
      full_path = self.add_triggerer_suffix(full_path=full_path, job_id=ti.triggerer_job.id)
  AttributeError: 'NoneType' object has no attribute 'id'
   
    warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
```

drajguru committed:
TriggererJob/TriggererJobRunner/Trigger tests logging fix

uranusjr commented on tests/jobs/test_triggerer_job.py:
```diff
@@ -51,6 +52,24 @@
+    Prevent console handler from getting bound to the trigger_job_runner when setup_queue_listener is called during tests
+    """
+
+    class QueueListener_(logging.handlers.QueueListener):
```
Does this class need to be defined in the function scope?

drajguru replied:
True, I just moved it out

drajguru committed:
Move QueueListener_ and fixture to module scope

drajguru committed:
Merge branch 'main' into triggerer_logging

o-nikolas committed:
Introduce mechanism to support multiple executor configuration (#37635)

* Introduce mechanism to support multiple executor configuration

This commit delivers an isolated component of AIP-61:
https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-61+Hybrid+Execution

It updates the executor configuration logic to allow configuring multiple
executors (and aliases per executor) in a list form (comma delimited), as
well as associated methods for core Airflow code to load the default executor
(first executor in the list) or any other configured executor.

Testing is included with the changes but not documentation, because the
feature is not yet complete and is currently disabled. User facing
documentation will be delivered when the entire AIP-61 feature is
released.

* Move caches to module and use reloads to empty them in tests

Executor data is now more aggressively cached and this was failing some tests
that are trying to override the executor conf. Moved the caches to the module
level and added some reload(executor_loader) calls to empty the caches for
certain tests that require it.

* Fix TestBaseSensor::test_prepare_for_execution

This unit test was broken to begin with and was working mostly by
chance. Updated the patch to be the correct method to mock and test
is passing again

* More reloads

* Comment grammar fix

* New code needed a reload of executor loader module

* Fix unit test is-->are

bolkedebruin committed:
Update ObjectStoragePath for universal_pathlib>=v0.2.2 (#37930)

This updates ObjectStoragePath to be compatible with universal_pathlib >= 0.2.1 which in turn makes it compatible with Python 3.12+.

---------

Co-authored-by: Andreas Poehlmann <ap--@users.noreply.github.com>

poorvirohidekar committed:
Fix D105 for mongo provider (#38016)

potiuk committed:
Update packaging to latest version in our build dependencies (#38018)

Packaging has been updated to 24.0 - and since our build dependencies
should be pinned for reproducibilty - we have to manually bumpt them
when some of the few build dependencies got upgraded.

shahar1 committed:
Fix bug in validate-operators-init pre-commit (#38014)

potiuk committed:
Upgrade UV to 0.1.17 (#38019)

The latest UV **just** released should fix the problem introduced
in 0.1.16 that disallowed non-approved PEP keys in pyproject.toml

This PR bumps 0.1.17 as the base version for UV

potiuk committed:
Remove constraint caching from image build (#38020)

Constraint caching added recently to the image caused some
unexpected side effects when conflicting constraints were added - as
old constraints cached in the image were not replaced even if main has
been updated. This PR removes caching from the image - so that building
the image will always use constraints from main.

alizmhdi committed:
Fix D105 checks for SSH provider (#38013)

Co-authored-by: mehdi.alizade <mehdi.alizade@cafebazaar.ir>

potiuk committed:
Add Python 3.12 support (#36755)

Finally after a number of dependency upgrades we seem to be able to
upgrade to Python 3.12 (pending universal_pathlib 0.2.0 conversion)

Several providers are excluded from being installed and wait for
Python 3.12, but it should not block Airlfow's general 3.12 support.

Co-authored-by: dirrao <dirisala.gopal@gmail.com>

potiuk committed:
Revert "Add Python 3.12 support (#36755)" (#38023)

This reverts commit 76dee0b5867580480d802765bad2dd61e3156bff.

Taragolis committed:
Move `tests/core/test_impersonation_tests` into the Quarantined (#38024)

potiuk committed:
Improve suffix handling for provider-generated dependencies (#38029)

In order to better address chicken-egg provider problem (in the
light of upcoming release of 2.9.0 and separated FAB provider, we
should make sure that providers released together with Airflow in
the same pre-release version, but also with another providers they
refer to, have the same suffix for version.

For example:

* FAB released with dev0 should refer to airflow >= x.y.z.dev0
* FAB released with b0 (normalized from beta0) should refer to
  airflow>=x.y.zb0
* Postgres released with rc1 should refer to common-sql >= x.y.zrc1

This will only work when the packages released are released together
in a single wave (which we should do now for Airlfow and FAB.

Later on - this will be less of a problem, because this is the only
time where we do not have the X.Y.Z version of Airflow released and
mandatory provider (FAB) has >=X.Y.Z. At the same time airflow has
fab set as mandatory provider so we need to make sure both of them
are released in the same wave.

By releasing them together, we solve the chicken-egg provider
problem for mandatory providers.

Taragolis committed:
Get rid of deprecated `pytest.warns(None)` usage (#38027)

poorvirohidekar committed:
Fix D105 for Apache Hive Provider (#38042)

* Fix D105 for Apache Hive Provider

* Fix D105 for Apache Hive Provider

VincentChantreau committed:
Use `AwsLogsHook` when fetching Glue job logs (#38010)

ephraimbuddy committed:
Chart: Default airflow version to 2.8.3 (#38036)

potiuk committed:
Exclude Cassandra for Python 3.12 (#38050)

We need to exclude Cassandra separately, in order to make sure
that cassandra drivers are not installed in the Python 3.12 image
when preparing it from the main branch.

Generally speaking the dependencies are cached from the main branch
when image is built, which means that the resulting image will still
have cassandra driver even if we remove it in final pyproject.toml.

The way how to get rid of it is two fold:

1) add the exclusion to main
2) after it is merged, increase the EPOCH number in the Dockerfile.ci
   in the PR that adds Python 3.12 compatibility

shahar1 committed:
Fix all `mypy-core` references to `mypy-airflow` (#38046)

Taragolis committed:
Avoid use of `assert` outside of the tests (#37718)

potiuk committed:
Add Python 3.12 support (#36755) (#38025)

Finally after a number of dependency upgrades we seem to be able to
upgrade to Python 3.12 (pending universal_pathlib 0.2.0 conversion)

Several providers are excluded from being installed and wait for
Python 3.12, but it should not block Airlfow's general 3.12 support.

Co-authored-by: dirrao <dirisala.gopal@gmail.com>
(cherry picked from commit 76dee0b5867580480d802765bad2dd61e3156bff)

potiuk committed:
Remove typo when determining if extra PROD image checks should be run (#38055)

potiuk committed:
Fix casing of PYTHON when looping over python list in CI (#38058)

This caused failures in Cache upload for PROD images.

potiuk committed:
Extract CI image building to separate workflow (#38057)

Similarly to #37865, we are extracting CI image buolding to a
reusable workflow. Previously we used composite action (and for
now it is still used by build-image workflow, but building CI images
in workflow should get better output and presentability with foldable
group of building images.

potiuk committed:
Limit prod-image extra checks to default python version (#38060)

The extra checks for prod image are only to check is the variants
we have for prod image can be built. There is no need to run it for
all python versions - default python version is enough

shahar1 committed:
Rename `SparkSqlOperator`'s field's name to comply with templated fields validation (#38045)

Taragolis committed:
Fix case if `SparkKubernetesOperator.application_file` is templated file (#38035)

* Fix case if `SparkKubernetesOperator.application_file` is templated file

* Simplify non-templated file test

* Change check sequential order

* Unique dag names

potiuk committed:
Exclude Python 3.12 for Databricks provider (#38070)

Databricks is excluded for Python 3.12 because running
databricks-sql-python imports on Python 3.12 Cause extremely long import
times: https://github.com/databricks/databricks-sql-python/issues/369
and until the problem is fixed, we exclude Python 3.12 for Databricks
provider

vincbeck committed:
Add back system test for AWS auth manager (#38044)

potiuk committed:
Restore "iffy" solution for skipping tasks for in-workflow build (#38077)

GitHub actions has strange and complex way of getting conditional
jobs. When you want a job to depend on another job to be running
when it is either successful of fully skipped, you have to write
a rather peculiar condition:

```
always() && !failure() && !cancelled()
```

However, it turned out that this condition does not propoagate to
to task that depends on the downstream tasks - for example if
"wait-for-ci-images" had this condition and preceding "build-ci-images"
was skipped, the "wait-for-ci-images" was happily running, but
any of the other jobs that depended on "wait-for-ci-images" has
been skipped :(. Simply - skipping tasks immediately propagates
to all the downstream tasks and we cannot easily mitigate it.

This means that we have to come back to a solution that
was implemented before #38057 - where instead of skipping the whole
job, we skip all the steps in this kob. That makes it a little more
clunky but makes the skipping status not propagate to downstream
jobs.

vincbeck committed:
Replace "Role" by "Group" in AWS auth manager (#38078)

potiuk committed:
Update documentation about recent image changes (#38085)

potiuk committed:
Properly exclude providers from serialization tests (#38074)

The serialization test imports example dags from providers and when
providers are excluded in a certain Python version, they should be
excluded. The previous exclusion had a bug, that did not properly
checked exclusions (it missed adding ROOT_FOLDER to exclusion patterns)

This PR fixes it.

bbovenzi committed:
Fix graph task state border color (#38084)

dstandish committed:
Handle re-serialization of pydantic models (#37855)

Previously we handled serialization of TaskInstance and DagRun, converting to the respective pydantic model. E.g. we give a TI and it comes out as TaskInstancePydantic.  But when we passed TaskInstancePydantic to the serializer, we did not encode it, so it hit the default "encoding" which is to simply stringify it.  With this change, TaskInstancePydantic will serialize to TaskInstancePydantic, etc.

uranusjr commented on tests/jobs/test_triggerer_job.py:
```diff
@@ -51,6 +52,24 @@
+        no_console_handlers = []
+        for h in handlers:
+            if h.name != "console":
+                no_console_handlers.append(h)
```
```suggestion
        no_console_handlers = (h for h in handlers if h.name != "console")
```

uranusjr commented on tests/models/test_trigger.py:
```diff
@@ -142,10 +142,12 @@ def test_submit_failure(session, create_task_instance):
     Tests that unassigned triggers of all appropriate states are assigned.
     """
+    # disable QueueListener to prevent logging during tests
+    monkeypatch.setattr("airflow.jobs.triggerer_job_runner.DISABLE_LISTENER", True)
```
I wonder if there’s a better way… perhaps adding an argument on `TriggererJobRunner` for this? Not entirely sure.

uranusjr commented on tests/jobs/test_triggerer_job.py:
```diff
@@ -398,8 +427,11 @@ def handle_events(self):
     job_runner = TriggererJob_(job)
     job_runner.trigger_runner = TriggerRunner_()
+    # set job_id so that it can be used by FileTaskHandler.add_triggerer_suffix
+    job_runner.trigger_runner.job_id = job.id
```
Why is this relevant?

uranusjr commented on tests/jobs/test_triggerer_job.py:
```diff
@@ -361,7 +390,7 @@ class TriggererJob_(TriggererJobRunner):
         def wait_for_runner_loop(self, runner_loop_count):
             for _ in range(30):
                 time.sleep(0.1)
-                if getattr(self.trigger_runner, "call_count", 0) >= runner_loop_count:
+                if getattr(self.trigger_runner, "loop_count", 0) >= runner_loop_count:
```
Is there a way to detect this if the attribute gets renamed?

potiuk committed:
Limit importlib_resources as it breaks pytest_rewrites (#38095)

New version of importlib_resources released today - 6.2.0 but also 6.3.0
released few hours later, break pytest_rewrite feature in rather
unexpected ways. Until the problem is diagnosed and solved we should
limit it. The issue is tracked in
https://github.com/python/importlib_resources/issues/299

bbovenzi committed:
Add try number to mapped instance (#38097)

potiuk committed:
Skip pydantic model round-trip when AIP-44 is disabled (#38103)

potiuk committed:
Upgrade uv to latest 0.18.0 version (#38101)

potiuk committed:
Bumping minimum version of zenpy to 0.40 (#38110)

The older versions had broken pyproject.toml and caused uv to
crash when trying --lowest-direct resolution (which we are going
to start using soon to test minimum versions of our dependencies)

bbovenzi committed:
Fix task instances list link (#38096)

potiuk committed:
Get rid of lowest-direct resolution in no-constraint upgrades with uv (#38109)

When we fall-back to no-constraint mode in uv, we used so far
lowest-direct resolution - but it could cause a really old set of
versions to be installed with some strange pyproject.toml errors
(for example zenpy==2.0.32 had a buggy pyproject.toml and it
caused editable local install to fail.

potiuk committed:
Bump Cryptography to > 39.0.0 (#38112)

We had a very old cryptography version specified as min requirement.
Bumping cryptography to 39.0.0 (released Jan 2023) should help with
better resolutions and should keep Airflow secure.

potiuk committed:
Optimize skipping building images for not "in-workflow" builds (#38091)

In case images are built in the "pull-request-target" workflow, we
should skip the builds "in-workflow" - but just "skipping" them is
no good because skip state propagates to downstream jobs - but this
leads - in naive implementation of build matrix to a number of
jobs that start and end immediately.

Instead we can simply use empty matrix for the builds - this will
make it much better optimized.

bbovenzi committed:
Add dataset_expression to grid dag details (#38121)

jscheffl committed:
Feature/add log grouping to UI (#38021)

* Add folding feature for log group to web UI

* Add folding feature for log groups, documentation

* Group logs from pre-task execution and post-execution logs

* Add state to the log block in grid view when task is running and logs are updated

* Spelling

* Review feedback

* Review feedback v2

* Enhance unicode arrow symbol compatability

potiuk committed:
Separate MyPy checks to a separate workflow (#38129)

potiuk committed:
Separate Helm jobs to separate workflow (#38126)

Continuing work from #37865 and #38057 we are extracting Helm chart
to a separate workflow.

bbovenzi committed:
Show if dag page filters are active (#38080)

* Show users when Dag page filters are active

* keep border color on hover

* Clear filter button only for baseDate,runType,runState

* Make number of runs separate from filters

drajguru committed:
Change loop to generator comprehension
